tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/ExerciseBuilder.java

```diff
@@ -0,0 +1,107 @@
+
+        for (File projectFile : projectFiles) {
+            prepareStubFile(projectFile);
+        }
+    }
+
+    private void prepareStubFile(File file) {
+
+        try {
+            List<String> lines = Files.readAllLines(file.toPath(), charset);
+            List<String> filteredLines = new ArrayList<>();
+            boolean skipLine = false;
+            for (String line : lines) {
+                if (line.contains(beginSolution)) {
+                    skipLine = true;
+                } else if (skipLine && line.contains(endSolution)) {
+                    skipLine = false;
+                } else if (line.contains(stubMarker)) {
+                    String start = line.substring(0, line.indexOf(stubMarker) - 1);
+                    String end = line.substring(line.indexOf(stubMarker) + stubMarker.length());
+                    filteredLines.add(start + end);
+                } else if (!skipLine && !line.contains(stubMarker)) {
+                    filteredLines.add(line);
+                }
+            }
+            Files.write(file.toPath(), filteredLines, charset);
+        } catch (IOException ex) {
+            throw new RuntimeException("Unexpected IOException, preparation of file {" + file.getAbsolutePath() + "} interrupted", ex);
+        }
+    }
+
+    private List<File> getFileList(File folder) {
+        if (!folder.isDirectory()) {
+            System.out.println("No files found");
+            return new ArrayList<>();
+        }
+        ArrayList<File> result = new ArrayList<>();
+        for (File file : folder.listFiles()) {
+            if (file.isDirectory()) {
+                result.addAll(getFileList(file));
+                continue;
+            }
+            result.add(file);
+        }
+        return result;
+    }
+
+    /**
+     * Prepares a presentable solution from the original.
+     *
+     * <p>
+     * Implements LanguagePlugin.preparaSolution
+     */
+    public void prepareSolution(Path path) {
+        File projectRoot = path.resolve(sourceFolderName).toFile();
+        List<File> projectFiles = getFileList(projectRoot);
+
+        for (File projectFile : projectFiles) {
+            prepareSolutionFile(projectFile);
+        }
+    }
+
+    private void prepareSolutionFile(File file) {
+        try {
+            List<String> lines = Files.readAllLines(file.toPath(), charset);
+            List<String> filteredLines = new ArrayList<>();
+            for (String line : lines) {
+                if (line.contains(beginSolution) || line.contains(endSolution) || line.contains(stubMarker)) {
+                    continue;
+                }
+                filteredLines.add(line);
+            }
+            Files.write(file.toPath(), filteredLines, charset);
+        } catch (IOException ex) {
+            throw new RuntimeException("Unexpected IOException, preparation of file {" + file.getAbsolutePath() + "} interrupted", ex);
```

Comment:
Maybe IllegalStateException? and we don't need to mention IOException here as it will be included to the stack trace as we have the ex as the second parameter...